src/Miki.Framework/Extension/MiddlewareExtensions.cs

```diff
@@ -0,0 +1,152 @@
+﻿﻿using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Miki.Framework.Hosting;
+
+namespace Miki.Framework
+{
+    using Predicate = Func<IContext, bool>;
+    using PredicateAsync = Func<IContext, ValueTask<bool>>;
+    
+    public static class MiddlewareExtensions
+    {
+        /// <summary>
+        /// Adds a middleware delegate defined in-line to the application's request pipeline.
+        /// </summary>
+        /// <param name="app">The <see cref="IBotApplicationBuilder"/> instance.</param>
+        /// <param name="middleware">A function that handles the request or calls the given next function.</param>
+        /// <returns>The <see cref="IBotApplicationBuilder"/> instance.</returns>
+        public static IBotApplicationBuilder Use(this IBotApplicationBuilder app, Func<IContext, Func<ValueTask>, ValueTask> middleware)
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+            
+            if (middleware == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+            
+            return app.Use(next =>
+            {
+                return context =>
+                {
+                    return middleware(context, () => next(context));
+                };
+            });
+        }
+        
+        /// <summary>
+        /// Adds a terminal middleware delegate to the application's request pipeline.
+        /// </summary>
+        /// <param name="app">The <see cref="IBotApplicationBuilder"/> instance.</param>
+        /// <param name="handler">A delegate that handles the request.</param>
+        public static void Run(this IBotApplicationBuilder app, MessageDelegate handler)
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+
+            if (handler == null)
+            {
+                throw new ArgumentNullException(nameof(handler));
+            }
+
+            app.Use(_ => handler);
+        }
+        
+        /// <summary>
+        /// Conditionally creates a branch in the request pipeline that is rejoined to the main pipeline.
+        /// </summary>
+        /// <param name="app"></param>
+        /// <param name="predicate">Invoked with the request environment to determine if the branch should be taken</param>
+        /// <param name="configuration">Configures a branch to take</param>
+        /// <returns></returns>
+        public static IBotApplicationBuilder UseWhen(this IBotApplicationBuilder app, Predicate predicate, Action<IBotApplicationBuilder> configuration)
+        {
+            if (app == null)
```

Comment:
App can't be null here right? Either way, you can use `Required<IBotApplicationBuilder>` to automate the process.

Reply:
> App can't be null here right?

Probably not, [this is how Microsoft implemented `UseWhen` as well](https://github.com/dotnet/aspnetcore/blob/425c196cba530b161b120a57af8f1dd513b96f67/src/Http/Http.Abstractions/src/Extensions/UseWhenExtensions.cs#L23-L28) but I can remove this.

`Required<IBotApplicationBuilder>` is not possible because it's an extension method. It would register the extension method to `Required<IBotApplicationBuilder>` instead of `IBotApplicationBuilder`.

Reply:
Fair enough, but the others could.